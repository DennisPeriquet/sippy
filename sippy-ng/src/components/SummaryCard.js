import { Box, Card, CardContent, Tooltip, Typography } from '@mui/material'
import { Doughnut } from 'react-chartjs-2'
import { Link } from 'react-router-dom'
import { makeStyles, useTheme } from '@mui/styles'
import { scale } from 'chroma-js'
import InfoIcon from '@mui/icons-material/Info'
import PropTypes from 'prop-types'
import React from 'react'

const useStyles = makeStyles({
  cardContent: {
    color: 'black',
    textAlign: 'center',
  },
  summaryCard: (props) => ({
    height: '100%',
  }),
})

/**
 * SummaryCard with a simple pie chart showing some data,
 * along with a caption. Used mainly by the ReleaseOverview
 * page.
 */
export default function SummaryCard(props) {
  const classes = useStyles(props)
  const theme = useTheme()

  const percent =
    ((props.success + props.flakes) /
      ((props.flakes || 0) + props.fail + props.success)) *
    100

  const colors = scale([
    theme.palette.error.light,
    theme.palette.warning.light,
    theme.palette.success.light,
  ]).domain([
    props.threshold.error,
    props.threshold.warning,
    props.threshold.success,
  ])

  const bgColor = colors(percent).hex()

  const labels = ['Pass', 'Flake', 'Fail']
  const data = [props.success, props.flakes, props.fail]
  const color = [
    theme.palette.success.dark,
    theme.palette.warning.dark,
    theme.palette.error.dark,
  ]

  let card = (
    <Card
      elevation={5}
      className={`${classes.summaryCard}`}
      style={{ backgroundColor: bgColor }}
    >
      <CardContent className={`${classes.cardContent}`}>
        <Typography variant="h6">
          {props.name}
          {props.tooltip ? <InfoIcon /> : ''}
        </Typography>
        <div align="center">
          <div style={{ width: '70%' }}>
            <Doughnut
              data={{
                labels: labels,
                datasets: [
                  {
                    label: props.name,
                    data: data,
                    borderColor: bgColor,
                    backgroundColor: color,
                  },
                ],
              }}
              options={{
                plugins: {
                  legend: {
                    display: false,
                  },
                },
                cutout: '60%',
              }}
            />
          </div>
          {props.caption}
        </div>
      </CardContent>
    </Card>
  )

  // Wrap in tooltip if we have one
  if (props.tooltip !== undefined) {
    card = (
      <Tooltip title={props.tooltip} placement="top">
        {card}
      </Tooltip>
    )
  }

  // Link if we have one
  if (props.link !== undefined) {
    return (
      <Box component={Link} to={props.link}>
        {card}
      </Box>
    )
  } else {
    return card
  }
}

SummaryCard.defaultProps = {
  success: 0,
  fail: 0,
  flakes: 0,
  caption: '',
  tooltip: '',
  units: 'percent',
}

SummaryCard.propTypes = {
  units: PropTypes.string,
  flakes: PropTypes.number,
  success: PropTypes.number,
  fail: PropTypes.number,
  caption: PropTypes.oneOfType([PropTypes.object, PropTypes.string]),
  tooltip: PropTypes.string,
  name: PropTypes.string,
  link: PropTypes.string,
  threshold: PropTypes.shape({
    success: PropTypes.number,
    warning: PropTypes.number,
    error: PropTypes.number,
  }),
}
