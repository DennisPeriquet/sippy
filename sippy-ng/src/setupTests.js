/** @jest-environment setup-polly-jest/jest-environment-node */

import '@testing-library/jest-dom'
import Enzyme from 'enzyme'
import Adapter from '@wojtekmaj/enzyme-adapter-react-17'
import fetch from 'node-fetch'
import path from 'path'
import { setupPolly } from 'setup-polly-jest'
import toDiffableHtml from 'diffable-html'

// https://github.com/mui-org/material-ui/issues/21293
export const withoutMuiID = (wrapper) => toDiffableHtml(wrapper.html()
  .replace(/id="mui-[0-9]*"/g, '')
  .replace(/aria-labelledby="(mui-[0-9]* *)*"/g, ''))

Enzyme.configure({ adapter: new Adapter() })

// To stub out fetch(), we need to use node-fetch and this line:
global.fetch = fetch

// See: https://github.com/vuejs/vue-test-utils/issues/974
global.requestAnimationFrame = cb => cb()

// When injecting multiple filters, material table complains.
const originalError = console.error.bind(console.error)

global.console.error = (log) => {
  if (!log.toString().includes('Warning: Failed')) {
    originalError(log.toString())
  }
}

// Set API URL for Sippy
process.env.REACT_APP_API_URL = 'http://localhost:8080'

// Default PollyJS mode is replay, set to record to record
// new API calls.
if (!process.env.POLLY_MODE) {
  process.env.POLLY_MODE = 'replay'
}

export const setupDefaultPolly = () => {
  const context = setupPolly({
    mode: process.env.POLLY_MODE,
    adapters: [require('@pollyjs/adapter-node-http')],
    persister: require('@pollyjs/persister-fs'),
    persisterOptions: {
      fs: {
        recordingsDir: path.resolve(__dirname, '../__recordings__')
      }
    }
  })
  return context
}

export const expectLoadingPage = (wrapper) => expect(wrapper.find('p').contains('Loading...'))

jest.mock('react-chartjs-2', () => ({
  Bar: () => null,
  Chart: () => null,
  ChartComponent: () => null,
  Doughnut: () => null,
  Line: () => null,
  PolarArea: () => null
}))
