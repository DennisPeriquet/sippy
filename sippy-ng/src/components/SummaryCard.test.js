import React from 'react'
import { act } from '@testing-library/react'
import { withoutMuiID } from '../setupTests'
import { mount } from 'enzyme'
import SummaryCard from './SummaryCard'

jest.useRealTimers()

describe(SummaryCard, () => {
  it('shows overview', async () => {
    let wrapper
    await act(async () => {
      wrapper = mount(
          <SummaryCard flakes={10} success={10} fail={10} threshold={{ success: 90, warning: 80, error: 0 }} />
      )
    })

    expect(wrapper.find('ReactMinimalPieChart').exists()).toBe(true)
    expect(withoutMuiID(wrapper)).toMatchSnapshot()
  })
})
