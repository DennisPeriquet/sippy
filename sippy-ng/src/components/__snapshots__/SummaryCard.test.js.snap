// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SummaryCard shows overview 1`] = `
<div class="MuiPaper-root MuiCard-root h6">
  <div align="center">
    <div style="width: 70%;">
    </div>
  </div>
</div>
`;
